### DISCUSSION TITLE
Mitsubishi Electric Integration Missing Mode

### DISCUSSION THREAD

**Discussion by maslyankov** [arendst/Tasmota#10937] [Ideas]
### PROBLEM DESCRIPTION
I am using the Mitsubishi Electric module and it is working fine - turns on the AC to each mode, sets temp, etc, but it does not show the current mode the AC is in, it only shows ON or OFF.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
       #9237
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): NodeMCU
- [x] Tasmota binary firmware version number used: 9.1.0
  - [ ] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: espflashtool
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
19:19:56 CMD: Backlog Template; Module; GPIO 255
19:19:57 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:25:57","MiElHVAC":{"Temperature":19.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"030000090001A6000000000000000000","status":"06000000000000000000000000000000","stage":"09000000000000000000000000000000"},"TempUnit":"C"}
19:19:57 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:26:27","MiElHVAC":{"Temperature":19.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"030000090001A6000000000000000000","status":"06000000000000000000000000000000","stage":"09000000000000000000000000000000"},"TempUnit":"C"}

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
19:20:27 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:20:27","MiElHVAC":{"Temperature":19.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"030000090001A6000000000000000000","status":"06000000000000000000000000000000","stage":"09000000000000000000000000000000"},"TempUnit":"C"}

```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:

AC Model: Mitsubishi Electric MSZ-FH35VE
logs:
19:21:27 MQT: tele/tasmota_5F410B/HVACSETTINGS = {"Power":"ON","Temp":24.0,"FanSpeed":"auto","SwingV":"swing","SwingH":"RR","Bytes":"0200000109070007000005B000000000"}
19:21:29 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:21:29","MiElHVAC":{"Temperature":19.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000009006CA6000000000000000000","status":"06000006010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:21:37 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:21:37","MiElHVAC":{"Temperature":19.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000009006CA6000000000000000000","status":"06000007010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:21:47 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:21:47","MiElHVAC":{"Temperature":19.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000009006CA6000000000000000000","status":"06000008010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:21:55 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:21:55","MiElHVAC":{"Temperature":18.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000008006CA5000000000000000000","status":"06000008010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:21:56 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:21:56","MiElHVAC":{"Temperature":18.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000008006CA5000000000000000000","status":"06000009010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:21:57 MQT: tele/tasmota_5F410B/STATE = {"Time":"1970-01-01T19:21:57","Uptime":"0T19:21:59","UptimeSec":69719,"Heap":28,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"wifiname","BSSId":"77:DD:88:00:DD:EE","Channel":11,"RSSI":34,"Signal":-83,"LinkCount":1,"Downtime":"0T00:00:03"}}
19:21:57 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:21:57","MiElHVAC":{"Temperature":18.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000008006CA5000000000000000000","status":"06000009010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:21:57 MQT: tele/tasmota_5F410B/HVACSETTINGS = {"Power":"ON","Temp":24.0,"FanSpeed":"auto","SwingV":"swing","SwingH":"RR","Bytes":"0200000109070007000005B000000000"}
19:22:07 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:22:07","MiElHVAC":{"Temperature":18.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000008006EA5000000000000000000","status":"06000009010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:22:11 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:22:11","MiElHVAC":{"Temperature":18.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000008006EA5000000000000000000","status":"0600000A010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:22:12 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:22:12","MiElHVAC":{"Temperature":18.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000008006CA5000000000000000000","status":"0600000A010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:22:27 MQT: tele/tasmota_5F410B/STATE = {"Time":"1970-01-01T19:22:27","Uptime":"0T19:22:29","UptimeSec":69749,"Heap":28,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"wifiname","BSSId":"77:DD:88:00:DD:EE","Channel":11,"RSSI":32,"Signal":-84,"LinkCount":1,"Downtime":"0T00:00:03"}}
19:22:27 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:22:27","MiElHVAC":{"Temperature":18.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000008006CA5000000000000000000","status":"0600000A010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
19:22:27 MQT: tele/tasmota_5F410B/HVACSETTINGS = {"Power":"ON","Temp":24.0,"FanSpeed":"auto","SwingV":"swing","SwingH":"RR","Bytes":"0200000109070007000005B000000000"}
19:22:45 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-01T19:22:45","MiElHVAC":{"Temperature":18.0,"Operation":"ON","Compressor":"ON","roomtemp":"03000008006EA5000000000000000000","status":"0600000A010000000000000000000000","stage":"09000004010000000000000000000000"},"TempUnit":"C"}
```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
I have another ESP8266 running the same binary connected to a Mitsubishi Electric MSZ-SF50VE3 with no issues.
Working AC logs:
```
20:00:24 MQT: tele/tasmota_B0FCB0/HVACSETTINGS = {"Power":"ON","Mode":"heat","HAMode":"heat","Temp":23.0,"FanSpeed":"auto","SwingV":"5","SwingH":"C","Bytes":"0200000101080005000003AE00000000"}
20:01:08 MQT: tele/tasmota_B0FCB0/SENSOR = {"Time":"1970-02-06T20:01:08","MiElHVAC":{"Temperature":22.0,"Operation":"ON","Compressor":"ON","roomtemp":"0300000C0000AD000000000000000000","status":"06000012010000000000000000000000","stage":"09000000020000000000000000000000"},"TempUnit":"C"}
20:01:19 MQT: tele/tasmota_B0FCB0/SENSOR = {"Time":"1970-02-06T20:01:19","MiElHVAC":{"Temperature":22.0,"Operation":"ON","Compressor":"ON","roomtemp":"0300000C0000AD000000000000000000","status":"06000013010000000000000000000000","stage":"09000000020000000000000000000000"},"TempUnit":"C"}
```

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
![Screen Shot 2021-01-18 at 13 56 02](https://user-images.githubusercontent.com/12431631/104913353-25424a80-5996-11eb-8df9-6c0410039bb0.png)

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by smartboart** [comment]
I have the same issue with MSZEF25V....did you have found a solution for it?
I do not want to run Mitsubishi2mqtt because i already  run a lot of tasmota firmware here..

**Reply by deku-m** [reply]
Did anyone solved this? 
How did you configure it to use MITSUBISHI?

**Reply by maslyankov** [reply]
I am still experiencing the issue :(

**Comment by deku-m** [comment]
Mmm i just wanted to ask which version you have. As i like the tasmota
firmware but the library isnt included in 9.3 😬

**Reply by maslyankov** [reply]
The Mitsubishi Electric library is not included in the standard Tasmota builds. You need to compile Tasmota yourself with a small change in the user_config_override.h file. I have on hand such version of Tasmota 9.1 with enabled MIEL driver -> https://github.com/maslyankov/Tasmota-MIEL-HVAC
You can use that if it is easier. :)
edit: I uploaded the pre-built firmware to GitHub along with Home Assistant config file. :)

**Reply by deku-m** [reply]
Youre awesome thx lacking the skills yet for building it.

**Reply by moccajava** [reply]
> The Mitsubishi Electric library is not included in the standard Tasmota builds. You need to compile Tasmota yourself with a small change in the user_config_override.h file. I have on hand such version of Tasmota 9.1 with enabled MIEL driver -> https://github.com/maslyankov/Tasmota-MIEL-HVAC You can use that if it is easier. :) edit: I uploaded the pre-built firmware to GitHub along with Home Assistant config file. :)

may i know if the recent firmware releases already include the mitsubishi aircon library?

**Reply by sfromis** [reply]
The library has been included in all releases since october 2020, as an optional feature (not in standard builds), which you can enable in a custom build with `#define USE_MIEL_HVAC`

Doing a custom build is not a difficult thing to do, the tool TasmoCompiler makes it a few steps in a browser, without having downloaded any tools.
https://tasmota.github.io/docs/Compile-your-build/

**Reply by moccajava** [reply]
> The library has been included in all releases since october 2020, as an optional feature (not in standard builds), which you can enable in a custom build with `#define USE_MIEL_HVAC`
> 
> Doing a custom build is not a difficult thing to do, the tool TasmoCompiler makes it a few steps in a browser, without having downloaded any tools. https://tasmota.github.io/docs/Compile-your-build/

Thanks for the reply! Just wanted to check if it was already in the precompiled releases “tasmota.bin”, but yeah not an issue to do up the custom build. Will just need to keep doing it for newer releases.

thanks again!

**Reply by sfromis** [reply]
https://tasmota.github.io/docs/BUILDS/ shows features included with the basic binaries supplied.

**Comment by dgwynne** [comment]
It looks like your air conditioner has a new mode it operates in. The currently understood modes are these:

```c
#define MIEL_HVAC_UPDATE_MODE_HEAT	0x01
#define MIEL_HVAC_UPDATE_MODE_DRY	0x02
#define MIEL_HVAC_UPDATE_MODE_COOL	0x03
#define MIEL_HVAC_UPDATE_MODE_FAN	0x07
#define MIEL_HVAC_UPDATE_MODE_AUTO	0x08
```

But your unit is reporting 0x09. Do you have any idea what mode you think you're running in and what it's called?

**Reply by smartboart** [reply]
Maybe idle

**Reply by dgwynne** [reply]
Is that a mode you can set it to?

**Reply by smartboart** [reply]
I don't know it but I think not.. . I think it is just a information about the state of the machine.. Temperature rechaed.. no heating no cooling just idle.. . But it is just an idea... Not a fact..

**Reply by dgwynne** [reply]
I think I see what's happening. You've got i-See on that unit and it is enabled?

**Reply by grzegorz914** [reply]
> It looks like your air conditioner has a new mode it operates in. The currently understood modes are these:
> 
> ```c
> #define MIEL_HVAC_UPDATE_MODE_HEAT	0x01
> #define MIEL_HVAC_UPDATE_MODE_DRY	0x02
> #define MIEL_HVAC_UPDATE_MODE_COOL	0x03
> #define MIEL_HVAC_UPDATE_MODE_FAN	0x07
> #define MIEL_HVAC_UPDATE_MODE_AUTO	0x08
> ```
> 
> But your unit is reporting 0x09. Do you have any idea what mode you think you're running in and what it's called?

The new iSEE mode are:
9 = ISEE HEAT
10 = ISEE DRY
11 = ISEE COOL

**Comment by dgwynne** [comment]
Can you try the attached diff?

[miel-isee.txt](https://github.com/arendst/Tasmota/files/6449236/miel-isee.txt)

**Comment by maslyankov** [comment]
Here are the outputs for all the modes of the AC (Sorry for the delay):

`MiEl

Fan Only:

14:34:14 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-11T14:34:14","MiElHVAC":{"Temperature":23.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000D00ACAE000000000000000000","status":"06000000000000000000000000000000","stage":"09000000020000000000000000000000"},"TempUnit":"C"}
14:34:14 MQT: tele/tasmota_5F410B/HVACSETTINGS = {"Power":"ON","Mode":"fan_only","HAMode":"fan_only","Temp":24.0,"FanSpeed":"1","SwingV":"swing","SwingH":"RR","Bytes":"0200000107070207000005B000000000"}

Auto:

114:34:54 MQT: tele/tasmota_5F410B/HVACSETTINGS = {"Power":"ON","Mode":"auto","HAMode":"auto","Temp":23.0,"FanSpeed":"auto","SwingV":"auto","SwingH":"RR","Bytes":"0200000108080000000005AE00000000"}
14:34:54 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-11T14:34:54","MiElHVAC":{"Temperature":21.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000B00ACAA000000000000000000","status":"06000000000000000000000000000000","stage":"09000000020000000000000000000000"},"TempUnit":"C"}

Cool:

14:36:07 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-11T14:36:07","MiElHVAC":{"Temperature":23.0,"Operation":"ON","Compressor":"ON","roomtemp":"0300000D00ACAE000000000000000000","status":"06000006010000000000000000000000","stage":"09000000020000000000000000000000"},"TempUnit":"C"}
14:36:07 MQT: tele/tasmota_5F410B/HVACSETTINGS = {"Power":"ON","Temp":23.0,"FanSpeed":"1","SwingV":"swing","SwingH":"RR","Bytes":"020000010B080207000005AE00000000"}

Dry:

14:33:07 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-11T14:33:07","MiElHVAC":{"Temperature":23.0,"Operation":"ON","Compressor":"OFF","roomtemp":"0300000D00ACAE000000000000000000","status":"06000000010000000000000000000000","stage":"09000000060000000000000000000000"},"TempUnit":"C"}
14:33:07 MQT: tele/tasmota_5F410B/HVACSETTINGS = {"Power":"ON","Temp":23.0,"FanSpeed":"quiet","SwingV":"swing","SwingH":"RR","Bytes":"020000010A080107000005AE00000000"}

Heat:

14:33:37 MQT: tele/tasmota_5F410B/SENSOR = {"Time":"1970-01-11T14:33:37","MiElHVAC":{"Temperature":21.0,"Operation":"ON","Compressor":"OFF","roomtemp":"0300000B00ACAA000000000000000000","status":"06000000010000000000000000000000","stage":"09000000010000000000000000000000"},"TempUnit":"C"}
14:33:37 MQT: tele/tasmota_5F410B/HVACSETTINGS = {"Power":"ON","Temp":27.0,"FanSpeed":"quiet","SwingV":"swing","SwingH":"RR","Bytes":"0200000109040107000005B600000000"}

`

**Reply by dgwynne** [reply]
is that with the diff i provided above?

**Reply by maslyankov** [reply]
No. That is with the version from the first post above. Can I flash the fw with the diff OTA (gzipped I guess)?

**Reply by dgwynne** [reply]
I'll make you one...

I think  https://mild.embarrassm.net/~dlg/tasmota/tasmota-hvac.bin and https://mild.embarrassm.net/~dlg/tasmota/tasmota-hvac.bin.gz have the diff if you want to try them.

Their sha256 checksums are:

```
39facf54d6f20badf0abf523e63cf8f4cff5c4c3c47f1f904434dfd659b64c70  tasmota-hvac.bin
481783bcdc84a2b6c7f0781964c3126ba47cf80e08108ea572b4139cd5425900  tasmota-hvac.bin.gz
```

**Reply by maslyankov** [reply]
Flashing that (via OTA update) broke the fw of the device. After flashing it, any time there is any type of communication between the device and the hvac (ex, update temperature sensor, turn on ac etc), the device resets its settings:

```
0:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at FA, Count 121
00:00:00.054 FRC: Some settings have been reset (5)
00:00:00.059 Project tasmota MiEl-HVAC Version 9.4.0.4(hvac)-2_7_4_9(2021-05-17T21:01:41)
00:00:00.547 WIF: Connecting to AP1 Channel 1 BSSId  in mode 11n as tasmota_5F410B-0267...
00:00:01.753 WIF: Connected
00:00:02.003 HTP: Web server active on tasmota_5F410B-0267 with IP address 10.10.0.107
00:00:05.883 MQT: Attempting connection...
00:00:05.893 MQT: Connected
00:00:05.895 MQT: tele/tasmota_5F410B/LWT = Online (retained)
00:00:05.897 MQT: cmnd/tasmota_5F410B/POWER = 
00:00:05.908 MQT: tele/tasmota_5F410B/INFO1 = {"Info1":{"Module":"Generic","Version":"9.4.0.4(hvac)","FallbackTopic":"cmnd/DVES_5F410B_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:05.912 MQT: tele/tasmota_5F410B/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota_5F410B-0267","IPAddress":"10.10.0.107"}}
00:00:05.917 MQT: tele/tasmota_5F410B/INFO3 = {"Info3":{"RestartReason":{"Exception":28,"Reason":"Exception","EPC":["40236d3e","00000000","00000000"],"EXCVADDR":"00000008","DEPC":"00000000"}}}
00:00:06.029 QPC: Reset
00:00:10.061 MQT: tele/tasmota_5F410B/STATE = {"Time":"1970-01-01T00:00:10","Uptime":"0T00:00:11","UptimeSec":11,"Vcc":2.956,"Heap":30,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":674,"MqttCount":1,"Wifi":{"AP":1,"SSId":"","BSSId":"","Channel":1,"RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}
00:01:10.406 MQT: tele/tasmota_5F410B/STATE = {"Time":"1970-01-01T00:01:10","Uptime":"0T00:01:11","UptimeSec":71,"Vcc":2.979,"Heap":30,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":1,"Wifi":{"AP":1,"SSId":"","BSSId":"","Channel":1,"RSSI":52,"Signal":-74,"LinkCount":1,"Downtime":"0T00:00:03"}}
```

I'm not sure if this is caused by updating via OTA or by some incompatibility.

**Reply by hwebay** [reply]
since you used a self compiled firmware, you need to use another self compiled firmware for any firmware update in the future. If you use the OTA, you will get a precompiled firmware without the mitsubishi driver.
I am not really able to find the code to insert inside user_config_override.h :'( My google-fu is not strong enough

**Reply by maslyankov** [reply]
Refreshing device with above precompiled .bin did not solve the issue. Device reboots and resets settings due to exception after booting.

**Comment by brusarp** [comment]
Hi @dgwynne !
I'm also so pleased to use your integration!
My Mitsubishi unit is a MSZ-FD35VA Kirigamine with "Clean", "Plasma" and "I-See" functions.
I'm interested to help you to add this functions to your integration but from the bytes value I can't see any difference between the various power on command from IR remote...
How I can do that for you? Thank you

EDIT:
I try to set the weblog to 4... Here some results...

[Mitsubishi.txt](https://github.com/arendst/Tasmota/files/6749855/Mitsubishi.txt)

**Comment by grzegorz914** [comment]
Added in #22216
